### DISCUSSION TITLE
Request Example with multiple long running tasks in Lifecycle Events

### DISCUSSION THREAD

**Discussion by ssmails** [fastapi/fastapi#11260] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, APIRouter

Server = FastAPI(
    title=SERVICE_NAME,
    openapi_url="/v1/openapi.json",
    docs_url="/v1/docs")

root_router = APIRouter()

class BackgroundTasks(threading.Thread):
    def run(self, *args, **kwargs):
        kafka_consumer.consume()

@root_router.on_event("startup")
async def startup_event():
    # https://stackoverflow.com/questions/69886498/kafkaconsumer-prevents-other-asyncio-task-from-running-properly
    # https://stackoverflow.com/questions/70872276/fastapi-python-how-to-run-a-thread-in-the-background
    # todo update to use Lifecycle event (@asynccontextmanager)

    # 1. start ThreadPoolExecutor on cron schedule 
    thread = threading.Thread(target=start_worker_scheduler)
    thread.start()

    # 2. start kafka consumer
    # without running this as BackgroundTasks,
    # the Apis will not be served by server.
    tasks = BackgroundTasks()
    tasks.start()

Kafka Consume is running a forever loop 

    def consume(self):
        for message in self._consumer:
            do xxxx
```

### Description

I have a Fast Api server running with- 
1. Background task consuming messages from kafka topic (which must be started on server startup) and continue running until the server/application is stopped, so I put it in the @root_router.on_event("startup").
2. A scheduler task, creating a ThreadPoolExecutor with workers, that must also be started on server startup, and continue running until the server/application is stopped, so I put in in the @root_router.on_event("startup").

Question -
- How can I move this code to use Lifecycle events ?
- Are startup events going to be deprecated - if so, which version ?
- Is there any other alternative way to achieve this behavior outlined above?
- I could not create the scheduler thread as part of the background event (when I have the kafaka_consumer.consume() part of it. I had to explicitly have it in the startup event, after which it worked. Why is that ? 
- I could not create 2 blocking tasks in the startup event or the background task. Eg. kafka_consume and say another blocking task. Why ?

### Operating System

macOS

### FastAPI Version

fastapi~=0.110.0

### Pydantic Version

pydantic~=1.10.13

### Python Version

3.11

**Comment by jgould22** [comment]
Background tasks are really (in my opinion) best used for things like book keeping and logging outside of the controllers hot path so you can get the response to the client a bit quicker and tidy up after.

If you are using Gunicorn I would suggest you use [Gunicorn](https://docs.gunicorn.org/en/stable/settings.html?highlight=hooks#server-hooks) server hooks since they don't seem to be attached to any specific request.

**Reply by ssmails** [reply]
We are using 'import uvicorn', not Gunicorn. Is this possible without moving to Gunicorn or using a Gunicorn worker with Uvicorn ?
Also, if the Questions section above can be kindly answered, Thanks.

**Comment by YuriiMotov** [comment]
Let's close this for now to keep the discussions clean and organized.
If you’ve found a solution, please share it here.
If you still need help, feel free to let us know.
